docs/design-documents/20240812-recover-from-pipeline-errors.md

```diff
@@ -0,0 +1,194 @@
+before attempting to restart the pipeline, allowing time for issues to resolve
+themselves. The backoff parameters will be configurable to allow customization
+based on specific production needs.
+
+- **Configurable Parameters:**
+  - **Minimum Delay Before Restart:** Default: 1 second
+  - **Maximum Delay Before Restart:** Default: 1 minute
+  - **Backoff Factor:** Default: 2
+  - **Maximum Number of Retries:** Default: Infinite
+  - **Delay After Which Fail Count is Reset:** Default: 5 minutes (or on first
+    successfully end-to-end processed record)
+
+This results in a default delay progression of 1s, 2s, 4s, 8s, 16s, 32s, 1m,...,
+balancing the need for recovery time and minimizing downtime.
+
+### Consecutive failures and permanent stop
+
+To avoid endless restarts, Conduit will include an option to permanently stop
+the pipeline after a configurable number of consecutive failures. Once the
+pipeline successfully processes at least one record or runs without encountering
+an error for a predefined time frame, the count of consecutive failures will
+reset.
+
+## Pipeline state management
+
+A new pipeline state, `recovering`, will be introduced to indicate that a
+pipeline has experienced a fault and is in the process of recovering. This state
+will help differentiate between an actively running pipeline and one that is in
+a restart loop.
+
+- **State Transition:**
+  - `running` → `recovering` → `running` (after pipeline runs without failure
+    for the specified duration or a record is processed successfully end-to-end)
+  - `running` → `recovering` → `degraded` (if max retries reached or fatal error
+    occurs)
+
+This approach will improve the visibility of pipeline status and assist in
+debugging and monitoring.
+
+## Record deduplication
+
+Restarting a failed pipeline can result in some records being re-delivered, as
+some records might have already been processed by the destination, but the
+acknowledgments might not have reached the source before the failure occurred.
+For that case, we should add a mechanism in front of the destination which
+tracks records that have been successfully processed by the destination, but the
+acknowledgment wasn't processed by the source yet. If such a record is
+encountered after restarting the pipeline, Conduit should acknowledge it without
+sending it to the destination.
+
+## Audit log
+
+If a pipeline is automatically recovered, a user needs to check the Conduit logs
+to see the cause and what actually happened. Ideally that information would be
+retrievable through the API and would be attached to the pipeline itself so a
+user could get insights into how a pipeline is behaving.
+
+To add this feature, we would need to track what is happening on a pipeline and
+log any important events so we can return them to the user. When done properly,
+this could serve as an audit log of everything a pipeline went through (
+provisioning, starting, fault, recovery, stopping, deprovisioning etc.).
```

Comment:
I was mostly using it to illustrate what this could potentially look like. Although I think it would be nice to have more detailed state management, I don't think it's strictly needed for this specific feature.

Reply:
For the record, in one of the team syncs we talked about this and mentioned that we'll try to make all the new changes as generic as possible so they allow new pipeline states, but we won't introduce any other new pipeline states in the scope of the recovery epic.